### PR TITLE
Fixes test system and adds 1.19 test environment

### PR DESCRIPTION
**Pull request by TheLimeGlass**
### Description
- Fixes the skriptTestDev task from crashing and not allowing input from the console.
- Adds 1.19 to the test environment. All current tests run successfully.
- Adds `gradlew clean genDocs` command to generate the docs.
- Adds java version selection to the custom tasks.
- Sets the latest quickTest to 1.19 with Java 17.
- Fixes Java 8 tests on MC 1.13-1.16.
- Updates Gradle Wrapper to 7.4.2

### CODE REVIEW

**Comment by TPGamesNL** [comment]
You should change the quickTest in build.gradle to this test

**Comment by TheLimeGlass** [comment]
Oh ya, I also fixed the skriptTestDev task not allowing input from the console, that works again.

**Comment by TheLimeGlass** [comment]
Currently Paper errors because the implementation is 1.19 but the server is only Java 12 and doesn't support 1.19, so older versions still don't want to work.

File: .github/workflows/gradle.yml
Comment (skmedix) [inline .github/workflows/gradle.yml:10]: Let's bump all actions at once
```suggestion
            - uses: actions/checkout@v3
```

File: src/main/java/ch/njol/skript/effects/EffVisualEffect.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/effects/EffVisualEffect.java:49]: Accidentally included this

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
Looks good to me other than the one thing (as long as it all works properly)!

File: src/main/java/ch/njol/skript/effects/EffVisualEffect.java
Comment (TPGamesNL) [reply src/main/java/ch/njol/skript/effects/EffVisualEffect.java:49]: Probably fixed by merging master @TheLimeGlass

**Comment by TheLimeGlass** [comment]
Fixed it so now all tests from 1.13- 1.19 works. There is currently an issue in 1.13 which is causing the Java 12 test to fail, so not this pull request's fault.
```
[19:31:42 ERROR]: #!#! [Skript] Severe Error:
[19:31:42 ERROR]: #!#!
[19:31:42 ERROR]: #!#! Skript is running with developer command-line options.
[19:31:42 ERROR]: #!#! If you are not a developer, consider disabling them.
[19:31:42 ERROR]: #!#!
[19:31:42 ERROR]: #!#! Stack trace:
[19:31:42 ERROR]: #!#! java.lang.IllegalArgumentException: ItemStack must be a splash potion. This item stack was LINGERING_POTION.
[19:31:42 ERROR]: #!#!     at org.apache.commons.lang.Validate.isTrue(Validate.java:136)
[19:31:42 ERROR]: #!#!     at org.bukkit.craftbukkit.v1_13_R2.entity.CraftSplashPotion.setItem(CraftSplashPotion.java:24)
[19:31:42 ERROR]: #!#!     at ch.njol.skript.entity.ThrownPotionData.spawn(ThrownPotionData.java:116)
[19:31:42 ERROR]: #!#!     at ch.njol.skript.entity.ThrownPotionData.spawn(ThrownPotionData.java:45)
[19:31:42 ERROR]: #!#!     at ch.njol.skript.entity.EntityData.spawn(EntityData.java:417)
[19:31:42 ERROR]: #!#!     at ch.njol.skript.sections.EffSecSpawn.walk(EffSecSpawn.java:160)
[19:31:42 ERROR]: #!#!     at ch.njol.skript.lang.EffectSectionEffect.walk(EffectSectionEffect.java:44)
[19:31:42 ERROR]: #!#!     at ch.njol.skript.lang.TriggerItem.walk(TriggerItem.java:89)
[19:31:42 ERROR]: #!#!     at ch.njol.skript.lang.Trigger.execute(Trigger.java:56)
[19:31:42 ERROR]: #!#!     at ch.njol.skript.SkriptEventHandler.check(SkriptEventHandler.java:142)
[19:31:42 ERROR]: #!#!     at ch.njol.skript.SkriptEventHandler$PriorityListener.lambda$new$0(SkriptEventHandler.java:71)
[19:31:42 ERROR]: #!#!     at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:80)
[19:31:42 ERROR]: #!#!     at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:66)
[19:31:42 ERROR]: #!#!     at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:520)
[19:31:42 ERROR]: #!#!     at ch.njol.skript.Skript$1.lambda$run$0(Skript.java:633)
[19:31:42 ERROR]: #!#!     at org.bukkit.craftbukkit.v1_13_R2.scheduler.CraftTask.run(CraftTask.java:84)
[19:31:42 ERROR]: #!#!     at org.bukkit.craftbukkit.v1_13_R2.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:449)
[19:31:42 ERROR]: #!#!     at net.minecraft.server.v1_13_R2.MinecraftServer.b(MinecraftServer.java:1010)
[19:31:42 ERROR]: #!#!     at net.minecraft.server.v1_13_R2.DedicatedServer.b(DedicatedServer.java:439)
[19:31:42 ERROR]: #!#!     at net.minecraft.server.v1_13_R2.MinecraftServer.a(MinecraftServer.java:940)
[19:31:42 ERROR]: #!#!     at net.minecraft.server.v1_13_R2.MinecraftServer.run(MinecraftServer.java:837)
[19:31:42 ERROR]: #!#!     at java.base/java.lang.Thread.run(Thread.java:835)
[19:31:42 ERROR]: #!#!
[19:31:42 ERROR]: #!#! Version Information:
[19:31:42 ERROR]: #!#!   Skript: 2.6.2 (custom version)
[19:31:42 ERROR]: #!#!     Flavor: selfbuilt-unknown
[19:31:42 ERROR]: #!#!     Date: unknown
[19:31:42 ERROR]: #!#!   Bukkit: 1.13.2-R0.1-SNAPSHOT
[19:31:42 ERROR]: #!#!   Minecraft: 1.13.2
[19:31:42 ERROR]: #!#!   Java: 12.0.2 (OpenJDK 64-Bit Server VM 12.0.2+10)
[19:31:42 ERROR]: #!#!   OS: Windows 10 amd64 10.0
[19:31:42 ERROR]: #!#!
[19:31:42 ERROR]: #!#! Server platform: Paper
[19:31:42 ERROR]: #!#!
[19:31:42 ERROR]: #!#! Current node: null
[19:31:42 ERROR]: #!#! Current item: spawn [[entitytype:thrown lingering potions of health]] at {_l} (as org.bukkit.Location)
[19:31:42 ERROR]: #!#!
[19:31:42 ERROR]: #!#! Thread: Server thread
[19:31:42 ERROR]: #!#!
[19:31:42 ERROR]: #!#! Language: english
[19:31:42 ERROR]: #!#! Link parse mode: DISABLED
[19:31:42 ERROR]: #!#!
[19:31:42 ERROR]: #!#! End of Error.
```

**Comment by TheLimeGlass** [comment]
Completed! The Java 8 test is failing because of https://github.com/SkriptLang/Skript/issues/4897

File: .github/workflows/java-8-builds.yml
Comment (TPGamesNL) [inline .github/workflows/java-8-builds.yml:1]: While we could put these in the same workflow (skriptTestFull), I think it's not a bad idea to have them in separate workflows for clarity 👍

File: .github/workflows/java-17-builds.yml
Comment (TPGamesNL) [inline .github/workflows/java-17-builds.yml:20]: Why was this needed now, it had always worked without it

File: README.md
Comment (TPGamesNL) [inline README.md:67]:
```suggestion
./gradlew (quickTest|skriptTest|skriptTestJava8|skriptTestFull)
```
matches the order of the list below

File: README.md
Comment (TPGamesNL) [inline README.md:72]:
```suggestion
<code>skriptTest</code> (1.17+) runs the tests on the latest supported Java version.
<code>skriptTestJava8</code> (1.13-1.16) runs the tests on the oldest supported Java version.
```

**Review by AyhamAl-Ali — APPROVED** [review]
LGTM STICKER

File: build.gradle
Comment (TPGamesNL) [inline build.gradle:198]: Integer.parseInt returns a primitive `int`, or throws an exception if the value isn't correct. It can't be null

File: build.gradle
Comment (TPGamesNL) [inline build.gradle:196]: I think it'd be better to create constants for these Java versions, especially the J17 one. 

I'm also not sure about fixing the latest supported Java version, as Skript doesn't depend much on the Java version.

I think these directories (main and extra) should be renamed, as this repurposes them (the original idea was to have the main tests be a few tests of versions people were most likely to use)

File: build.gradle
Comment (TPGamesNL) [inline build.gradle:194]: There should be a constant for latest version instead of having to edit the build script IMO

**Review by TPGamesNL — CHANGES_REQUESTED** [review]

File: .github/workflows/java-17-builds.yml
Comment (TheLimeGlass) [reply .github/workflows/java-17-builds.yml:20]: I've had issues in the past where actions would error because of it. Maybe GitHub automatically applies it now? I'm not taking any chances.

File: build.gradle
Comment (TheLimeGlass) [reply build.gradle:198]: There are two lines to it, it's a null check for the property if the user forgot to add it.

**Comment by TheLimeGlass** [comment]
Added `gradlew clean genDocs` command to generate the docs as per @AyhamAl-Ali suggestion.

File: src/main/java/ch/njol/skript/SkriptCommand.java
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/SkriptCommand.java:88]:
```suggestion
		if (TestMode.GEN_DOCS || new File(Skript.getInstance().getDataFolder() + "/doc-templates").exists()) {
```

File: src/main/java/ch/njol/skript/SkriptCommand.java
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/SkriptCommand.java:93]: No need for this due to the above suggestion

File: src/main/java/ch/njol/skript/SkriptCommand.java
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/SkriptCommand.java:365]: Can you make it copy the `docs` folder contens from the source/jar to `plugins/Skript/doc-templates`

File: src/main/java/ch/njol/skript/tests/platform/Environment.java
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/tests/platform/Environment.java:230]: Also add `skript.forceregisterhooks` to force generating hooks without the need to install the plugins

File: src/main/java/ch/njol/skript/tests/platform/Environment.java
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/tests/platform/Environment.java:231]: We need a place to store the new docs as well

**Review by AyhamAl-Ali — CHANGES_REQUESTED** [review]
Amazing! Just little notes

File: src/main/java/ch/njol/skript/SkriptCommand.java
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/SkriptCommand.java:365]: No because you can't copy directories from the resource.

Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/SkriptCommand.java:365]: Once you add it once to the test runner folder, you don't have to do it again.

File: src/main/java/ch/njol/skript/tests/platform/Environment.java
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/tests/platform/Environment.java:231]: It goes into test_runner/ENV/plugins/Skript

File: src/main/java/ch/njol/skript/tests/platform/Environment.java
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/tests/platform/Environment.java:230]: Doesn't you pull request add this?

Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/tests/platform/Environment.java:230]: It doesn't, it only checks if this property is used to generate the hooks as well regardless of whether the plugins are intalled or not

File: src/main/java/ch/njol/skript/tests/platform/Environment.java
Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/tests/platform/Environment.java:231]: Okay cool then!

File: src/main/java/ch/njol/skript/SkriptCommand.java
Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/SkriptCommand.java:365]: The idea of having a gradle task for this is to generate the docs automatically with one click and no need to copy/paste anything manually and since the templates may change from time to time it wouldn't be ideal this way, I can do that in my PR if you want (add a check in `/sk gen-docs` if is SkriptTest.GEN_DOCS then copy the templates from the jar to Skript folder)

Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/SkriptCommand.java:365]: You can't without hard coding every file. You have to know every item in the docs folder, you can't transfer directories as an input stream is a single file.

Comment (TPGamesNL) [reply src/main/java/ch/njol/skript/SkriptCommand.java:365]: You can make the gradle task do it, doesn't have to use the resource system Skript has built in. Otherwise, you can extend that resource system to directories

Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/SkriptCommand.java:365]: I think it might be possible if we insert a .TXT file into the jar with a list of the file paths generated by Gradle when it puts the docs folder in the jar.

Comment (TPGamesNL) [reply src/main/java/ch/njol/skript/SkriptCommand.java:365]: Why can't we just make the gradle task copy everything from `docs` to `test-runners/runner/plugins/Skript/doc-templates`?

Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/SkriptCommand.java:365]: That's assuming we're not going to change the location, and also it can not be present, so it'll have to create the folders/files too. That is a solution.

If we put it in the jar, and make the map .txt solution, we could also use it for users that gen-docs with the commands

File: .github/workflows/java-8-builds.yml
Comment (APickledWalrus) [inline .github/workflows/java-8-builds.yml:30]: New line for end of file

File: src/main/java/ch/njol/skript/tests/TestResults.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/tests/TestResults.java:34]: Why not something like `docsFailed`?

Also goes for other occurrences of this name case.

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/tests/TestResults.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/tests/TestResults.java:34]: Field name should be changed here too

File: src/main/java/ch/njol/skript/tests/platform/PlatformMain.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/tests/platform/PlatformMain.java:89]: Field name should be changed here too

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
Sorry to be nitpicky about it lol - should've specified this directly in my original review

**Review by APickledWalrus — APPROVED** [review]
LGTM

Not sure if any discussion is still ongoing about the docs command stuff, but that should be resolved first of course.

**Review by AyhamAl-Ali — APPROVED** [review]
For now all looks good, the docs suggestions about copying files doesn't have to delay this, we can improve it in another PR.

**Comment by TheLimeGlass** [comment]
Waiting on review from @TPGamesNL

File: .github/workflows/java-8-builds.yml
Comment (TPGamesNL) [inline .github/workflows/java-8-builds.yml:24]: More an idea for the future, but I think it would be nice to have even more separation in our workflows, e.g. one workflow per MC version

File: build.gradle
Comment (TPGamesNL) [inline build.gradle:198]: Do you think it'd be better to rename this test skriptTestJava17 (to keep in line with naming of skriptTestJava8)?
(possibly also rename skriptTestFull to skriptTest after that)

File: src/main/java/ch/njol/skript/tests/platform/PlatformMain.java
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/tests/platform/PlatformMain.java:66]: Why make this parameter optional? It will only make things more annoying if we add more parameters in the future

File: src/main/java/ch/njol/skript/tests/platform/Environment.java
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/tests/platform/Environment.java:271]: Results shouldn't be read in DEV_MODE, right? When testing myself, even when using the `sk test` command, no results are written

File: src/main/java/ch/njol/skript/SkriptCommand.java
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/SkriptCommand.java:380]: When using the test command, e.g. `sk test syntaxes/conditions/CondContains`, and I run `sk test` again (or just the same test again with the full command) it runs that test multiple times, once for each command previously executed:
image

Can be fixed by unloading the script after running the tests

File: .github/workflows/java-8-builds.yml
Comment (TheLimeGlass) [reply .github/workflows/java-8-builds.yml:24]: I think you can only have 6 running at a time. Don't want to overload GitHub and get in trouble or having it wait on eachother. I get it.

File: .github/workflows/java-17-builds.yml
Comment (TheLimeGlass) [reply .github/workflows/java-17-builds.yml:14]: Sure, dependabot would catch this. GitHub doesn't have v3 labeled on the repository so that's why I didn't see.

File: build.gradle
Comment (TheLimeGlass) [reply build.gradle:198]: Sure

File: src/main/java/ch/njol/skript/tests/platform/PlatformMain.java
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/tests/platform/PlatformMain.java:66]: Sure

File: .github/workflows/java-8-builds.yml
Comment (TPGamesNL) [reply .github/workflows/java-8-builds.yml:24]: Looked it up, it's actually 20 concurrent jobs [[1]](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits) [[2]](https://github.com/github/docs/discussions/1724#discussioncomment-146788). Regardless, I've commented it at https://github.com/SkriptLang/Skript/issues/3767#issuecomment-1192691981 because it's out of scope of this PR

File: src/main/java/ch/njol/skript/SkriptCommand.java
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/SkriptCommand.java:380]: For some reason `ScriptLoader.disableScripts()` is commented out, so there may have been repercussions with doing this.

File: src/main/java/ch/njol/skript/tests/platform/Environment.java
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/tests/platform/Environment.java]: Oh I think I get what you mean. The file doesn't exist when in dev mode, so it throws an error when closing the server. Fixed.

File: src/main/java/ch/njol/skript/SkriptCommand.java
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/SkriptCommand.java:380]: Looks like you commented this code out lol.

Re-adding it back doesn't seem to affect anything.

File: .github/workflows/repo.yml
Comment (TPGamesNL) [inline .github/workflows/repo.yml:15]:
```suggestion
              uses: actions/setup-java@v3
```

File: build.gradle
Comment (TPGamesNL) [inline build.gradle:205]: Can genDocs be replaced with regular createTestTask now?

**Review by TPGamesNL — APPROVED** [review]

**Pull request merged**
